src/ShoppingList.js

```diff
@@ -0,0 +1,45 @@
+import React, { Component } from "react";
+import ListItems from './ListItems';
+
+class ShoppingList extends Component {
+  constructor(props) {
+    super(props);
+    this.addItem = this.addItem.bind(this);
+    this.state = { items: [] };
+  }
+
+  addItem(e) {
+    if (this._inputElement.value !== "") {
+      var newItem = {
+        text: this._inputElement.value,
+        key: Date.now()
+      };
+
+      this.setState((prevState) => {
+        return {
+          items: prevState.items.concat(newItem)
+        };
+      });
+
+      this._inputElement.value = "";
+    }
+    console.log(this.state.items);
+
+    e.preventDefault();
+  }
+
+  render() {
+    return (
+      <div>
+        <form onSubmit={this.addItem}>
+          <input ref={(a) => this._inputElement = a} placeholder="Item Name">
```

Comment:
I think this is probably the only part where I wouldn't describe it as 'idiomatic react'.

Instead of using a ref to the input element, you could add a second field (`text`) to the state that stores the value in the input box, so in your constructor the state would become:

```js
this.state = { items: [], text: '' };
```

Then to link the input element to the state you'd set the `value` and `onChange` properties:

```js
<input value={this.state.text} onChange={(e) => this.setState({ text: e.target.value })} placeholder="Item Name">
```

Then in `addItem()` to read the value of the text field you can use `this.state.text`, and to write the value you can call `this.setState()`:

```js
addItem(e) {
  if (this.state.text !== '') {
    var newItem = {
      text: this.state.text,
      key: Date.now()
    };

    this.setState((prevState) => {
      return {
        items: prevState.items.concat(newItem)
      };
    });

    this.setState({ text: '' });
  }
  console.log(this.state.items);

  e.preventDefault();
}
```

Avoiding refs for anything that can be done declaratively makes the code a lot more predictable, and means you don't have to worry about interacting with the dom elements directly in your callback function. It also makes debugging a lot easier if something isn't working as you can see the value of `this.state.text`.

Reply:
This is really useful, thanks, it's clear I need to spend some more time on react as we go with this project!